lib/modules.nix

```diff
@@ -412,10 +416,30 @@ rec {
       # modules recursively. It returns the final list of unique-by-key modules
       filterModules = modulesPath: { disabled, modules }:
         let
-          moduleKey = m: if isString m && (builtins.substring 0 1 m != "/")
-            then toString modulesPath + "/" + m
-            else toString m;
-          disabledKeys = map moduleKey disabled;
+          moduleKey = file: m:
+            if isString m
+            then
+              if builtins.substring 0 1 m == "/"
+              then m
+              else toString modulesPath + "/" + m
+
+            else if isConvertibleWithToString m
+            then
+              if m?key
+              then
+                throw "Module `${file}` contains a disabledModules item that is an attribute set that can be converted to a string (${toString m}) but also has a `.key` attribute (${m.key}) with a different value. This makes it ambigous which module should be disabled."
```

Comment:
```suggestion
            else if isConvertibleWithToString m
            then
              if m ? key && m.key != toString m
              then
                throw "Module `${file}` contains a disabledModules item that is an attribute set that can be converted to a string (${toString m}) but also has a `.key` attribute (${m.key}) with a different value. This makes it ambigous which module should be disabled."
```

Reply:
Done.
The other error messages are already ok
 - the one for non-string attrsets without key should keep recommending only the most explicit option: `.key`. It is also _technically correct_, even if it ignores the possibility of fixing the problem by making it `toString`-able. I'd like to keep the message simple and not explain what I consider irrelevant legacy behavior. If you disagree please leave a suggestion.
 - the final catch-all one already lists `value supported by toString`.